examples/step-19/doc/results.dox

```diff
@@ -0,0 +1,290 @@
+      auto end_particle_on_cell = begin_particle_on_cell;
+      while (end_particle_on_cell->get_surrounding_cell(triangulation)
+             == cell)
+        {
+          ++n_particles_on_cell;
+          ++end_particle_on_cell;
+        }
+
+      ...now operate on the range of particles from begin_particle_on_cell
+         to end_particle_on_cell, all of which are known to be on the current
+         cell...;
+
+      // Move the begin iterator forward so that it points to the first
+      // particle on the next cell
+      begin_particle_on_cell = end_particle_on_cell;
+    }
+@endcode
+
+In this code, we touch every cell exactly once and we never have to search
+the big data structure for the first or last particle on each cell. As a
+consequence, the algorithm costs a total of ${\cal O}(N)$ for a complete
+sweep of all particles and all cells.
+
+It would not be very difficult to implement this scheme for all three of the
+functions in this program that have this issue.
+
+
+<h4> More statistics about electrons </h4>
+
+The program already computes the fraction of the electrons that leave the
+domain through the hole in the anode. But there are other quantities one might be
+interested in. For example, the average velocity of these particles. It would
+not be very difficult to obtain each particle's velocity from its properties,
+in the same way as we do in the `move_particles()` function, and compute
+statistics from it.
+
+
+<h4> A better-synchronized visualization </h4>
+
+As discussed above, there is a varying time difference between different frames
+of the video because we create output for every time step. A better way to
+create movies would be to generate a new output file in fixed time intervals,
+regardless of how many time steps lie between each such point.
+
+
+<h4> A better time integrator </h4>
+
+The problem we are considering in this program is a coupled, multiphysics
+problem. But the way we solve it is by first computing the (electric) potential
+field and then update the particle locations. This is what is called an
+"operator-splitting method", a concept we will investigate in more detail
+in step-58.
+
+While it is awkward to think of a way to solve this problem that does not involve
+splitting the problem into a PDE piece and a particles piece, one
+*can* (and probably should!) think of a better way to update the particle
+locations. Specifically, the equations we use to update the particle location
+are
+@f{align*}{
+  \frac{{\mathbf v}_i^{(n)}-{\mathbf v}_i^{(n-1)}}{\Delta t} &= \frac{e\nabla V^{(n)}}{m}
+  \\
+  \frac{{\mathbf x}_i^{(n)}-{\mathbf x}_i^{(n-1)}}{\Delta t} &= {\mathbf v}_i^{(n)}.
+@f}
+This corresponds to a simple forward-Euler time discretization -- a method of
+first order accuracy in the time step size $\Delta t$ that we know we should
+avoid because we can do better. Rather, one might want to consider a scheme such
+as the leapfrog scheme, or even better a Runge-Kutta integrator.
```

Comment:
Tiny comments. For this type of problem (where the force/acceleration is only a function of the position), classical Verlet integrator (such as velocity-verlet or leapfrog)  are symplectic (https://en.wikipedia.org/wiki/Symplectic_integrator). This gives amazing properties, notably in terms of preserving the energy of the particles.  Consequently, they are quite desirable even if they are of lower order (hence their extensive use in molecular dynamics simulations). I'm not sure I would prefer classical RK over a velocity-verlet scheme here.

Reply:
Most excellent idea!